servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/NettyPipelineSslUtils.java

```diff
@@ -65,25 +116,34 @@ public static SSLSession extractSslSessionAndReport(ChannelPipeline pipeline,
                                                         SslHandshakeCompletionEvent sslEvent,
                                                         Consumer<Throwable> failureConsumer,
                                                         boolean shouldReport) {
-        final SecurityHandshakeObserver observer = shouldReport ? handshakeObserver(pipeline) : null;
+        final SecurityHandshakeObserver observer = lookForHandshakeObserver(pipeline, shouldReport);
         if (sslEvent.isSuccess()) {
             final SslHandler sslHandler = pipeline.get(SslHandler.class);
             if (sslHandler != null) {
-                final SSLSession session = sslHandler.engine().getSession();
-                if (observer != null) {
-                    observer.handshakeComplete(session);
-                }
-                return session;
+                return getSslSession(sslHandler, observer);
             } else {
-                deliverFailureCause(failureConsumer, new IllegalStateException("Unable to find " +
-                        SslHandler.class.getName() + " in the pipeline."), observer);
+                deliverFailureCause(failureConsumer, unableToFindSslHandler(), observer);
             }
         } else {
             deliverFailureCause(failureConsumer, sslEvent.cause(), observer);
         }
         return null;
     }
 
+    private static boolean noSslHandlers(final ChannelPipeline pipeline) {
+        return pipeline.get(SslHandler.class) == null && pipeline.get(DeferSslHandler.class) == null &&
```

Comment:
side note: would it make sense to give all of these handlers a common name so that they are easier to locate by name instead of their class type?

Reply:
That's a good idea that we can keep for future enhancements. The only one potential problem is that sometimes some name is hardcoded outside of ST control. See `io.netty.handler.ssl.SniHandler.replaceHandler` as an example.